### PR TITLE
[Core] Handle parent return/assign from FuncCall with No Scope

### PR DESCRIPTION
**Pull request by samsonasik**
Continue of https://github.com/rectorphp/rector-src/pull/2523, given the following code:

```php
trait ServiceLocatorTrait
{
    private array $factories;
    private array $loading = [];
    private array $providedTypes;

    public function get(string $id): mixed
    {
        if (!isset($this->factories[$id])) {
            throw $this->createNotFoundException($id);
        }

        $this->loading[$id] = $id;
        try {
            return $this->factories[$id]($this);
        } finally {
            unset($this->loading[$id]);
        }
    }
}
```

cause error:

```bash
php ../e2eTestRunner.php
 1/1 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓] 100%[ERROR] Could not process "src/ServiceLocatorTrait.php" file, due to:
         "System error: "Scope not available on "PhpParser\Node\Expr\FuncCall" node with parent node of
         "PhpParser\Node\Stmt\Return_", but is required by a refactorWithScope() method of
         "Rector\DowngradePhp81\Rector\FuncCall\DowngradeArrayIsListRector" rule. Fix scope refresh on changed nodes
         first"
         Run Rector with "--debug" option and post the report here: https://github.com/rectorphp/rector/issues/new". On
         line: 43%    
```

### CODE REVIEW

**Comment by samsonasik** [comment]
Also added multi assign on after Unreachable stmt https://github.com/rectorphp/rector-src/pull/2525/files#diff-2b611fae166e1b8d296735b6264da10ef54b94ab79e0158ea2baea0bcdda5056 that cause error:

```bash
Error: R] Could not process "src/MultipleAssign.php" file, due to:
-         "System error: "Scope not available on "PhpParser\Node\Expr\FuncCall"
-         node with parent node of "PhpParser\Node\Expr\Assign", but is required
-         by a refactorWithScope() method of
-         "Rector\DowngradePhp81\Rector\FuncCall\DowngradeArrayIsListRector"
-         rule. Fix scope refresh on changed nodes first"
-         Run Rector with "--debug" option and post the report here:
-         https://github.com/rectorphp/rector/issues/new". On line: 43
```

**Comment by samsonasik** [comment]
Progress:

- Fixed return function no scope part by verify if no previous node, lookup parent if parent is unreachable already https://github.com/rectorphp/rector-src/pull/2525/commits/f8451feb51e31149970ad8488c23961242a4cf71

**Comment by samsonasik** [comment]
I moved init scope for on before `refactor()` called from `AbstractScopeAwareRector` to `AbstractRector` so can fill scope for unreachable stmt with indirect Scope usage in the any rector rule https://github.com/rectorphp/rector-src/pull/2525/commits/bd71735d7ffbb0f71e5a5d3933de68e1e3344625

**Comment by samsonasik** [comment]
Fixed 🎉

**Comment by samsonasik** [comment]
I tried downgrade script, and it got different error:

```bash
3439/3439 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓] 100%
                                                                                                                        
 [ERROR] Could not process "rector-build/vendor/react/event-loop/src/StreamSelectLoop.php" file, due to:                
         "System error: "Scope not available on "PhpParser\Node\Expr\FuncCall" node with parent node of                 
         "PhpParser\Node\Stmt\Expression", but is required by a refactorWithScope() method of                           
         "Rector\DowngradePhp81\Rector\FuncCall\DowngradeArrayIsListRector" rule. Fix scope refresh on changed nodes    
         first"                                                                                                         
         Run Rector with "--debug" option and post the report here: https://github.com/rectorphp/rector/issues/new". On 
         line: 36                                                                                                       
                                                                                                                        

                                                                                                                        
 [ERROR] Could not process "rector-build/vendor/react/promise/src/Promise.php" file, due to:                            
         "System error: "Scope not available on "PhpParser\Node\Expr\FuncCall" node with parent node of                 
         "PhpParser\Node\Stmt\Expression", but is required by a refactorWithScope() method of                           
         "Rector\DowngradePhp81\Rector\FuncCall\DowngradeArrayIsListRector" rule. Fix scope refresh on changed nodes    
         first"                                                                                                         
         Run Rector with "--debug" option and post the report here: https://github.com/rectorphp/rector/issues/new". On 
         line: 36                                                                                                       
                                                                                                                        

                                                                                                                        
 [ERROR] Could not process "rector-build/vendor/symfony/dependency-injection/Argument/RewindableGenerator.php" file, due
         to:                                                                                                            
         "System error: "Node "PhpParser\Node\Expr\StaticCall" with parent of "PhpParser\Node\Expr\Ternary" is missing  
         scope required for scope refresh."                                                                             
         Run Rector with "--debug" option and post the report here: https://github.com/rectorphp/rector/issues/new". On 
         line: 72                                                                                                       
                                                                                                                        

                                                                                                                        
 [ERROR] Could not process "rector-build/vendor/symfony/console/Question/Question.php" file, due to:                    
         "System error: "Node "PhpParser\Node\Expr\StaticCall" with parent of "PhpParser\Node\Expr\Ternary" is missing  
         scope required for scope refresh."                                                                             
         Run Rector with "--debug" option and post the report here: https://github.com/rectorphp/rector/issues/new". On 
         line: 72                                                                                                       
                                                                                                                        

                                                                                                                        
 [ERROR] Could not process "rector-build/vendor/symfony/dependency-injection/Container.php" file, due to:               
         "System error: "Node "PhpParser\Node\Expr\StaticCall" with parent of "PhpParser\Node\Expr\BinaryOp\Coalesce" is
         missing scope required for scope refresh."                                                                     
         Run Rector with "--debug" option and post the report here: https://github.com/rectorphp/rector/issues/new". On 
         line: 72                                                                                                       
                                                                                                                        

                                                                                                                        
 [ERROR] Could not process "rector-build/vendor/symfony/finder/Iterator/SortableIterator.php" file, due to:             
         "System error: "Node "PhpParser\Node\Expr\StaticCall" with parent of "PhpParser\Node\Expr\Ternary" is missing  
         scope required for scope refresh."                                                                             
         Run Rector with "--debug" option and post the report here: https://github.com/rectorphp/rector/issues/new". On 
         line: 72                                                                                                       
                                                                                                     
```

I will continue searching.

**Comment by samsonasik** [comment]
I guess the test can be moved to `rules-tests` since it mostly issue on single rule `DowngradeArrayIsListRector` without dependency

**Comment by samsonasik** [comment]
I moved tests from e2e to rules-tests https://github.com/rectorphp/rector-src/pull/2525/commits/fc6fbebb4e8d9d767e32e2a20bf107ee2a1bd131

**Comment by samsonasik** [comment]
I [added failing test case for crash FuncCall has parent Expression](https://github.com/rectorphp/rector-src/pull/2525/commits/50d952e4c5b099a750a6a888aba4063a6c5e7533)

**Comment by samsonasik** [comment]
Fixed 🎉 with check against StmtsAwareInterface https://github.com/rectorphp/rector-src/pull/2525/commits/fc383075b976d0e7856eccf49d299a329bddd093#diff-625e519d3d496a50287cab05a2ffb79a95877d1850083f9a90fdd74958a01020

**Comment by samsonasik** [comment]
it seems cause other test error, reverting https://github.com/rectorphp/rector-src/pull/2525/commits/7530bb2fd674eafdea9d48319186eee3133ae7af

**Comment by samsonasik** [comment]
solved with [try verify with Scope exists](https://github.com/rectorphp/rector-src/pull/2525/commits/c72d1fac9af55cc04a9d9dc19ff44bd511bbd6ee)

**Comment by samsonasik** [comment]
There is still another error to check on downgrade script run:

```bash
 bin/rector process vendor/symfony/dependency-injection/Container.php --dry-run --config build/config/config-downgrade.php --clear-cache
 1/1 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓] 100%
                                                                                                                        
 [ERROR] Could not process "vendor/symfony/dependency-injection/Container.php" file, due to:                            
         "System error: "Node "PhpParser\Node\Expr\StaticCall" with parent of "PhpParser\Node\Expr\BinaryOp\Coalesce" is
         missing scope required for scope refresh."                                                                     
         Run Rector with "--debug" option and post the report here: https://github.com/rectorphp/rector/issues/new". On 
         line: 72                                                                                                       
                                                             
```

I will continue checking.

**Comment by samsonasik** [comment]
I added [failing test case parent Static Call from Coalesce](https://github.com/rectorphp/rector-src/pull/2525/commits/d17b641870e9053bdd6feccc0cb814985a06d140) for `DowngradeFirstClassCallableSyntaxRector`

**Comment by samsonasik** [comment]
Fixed 🎉

**Comment by samsonasik** [comment]
I tried downgrade script and it seems working ok 🎉 

Screen Shot 2022-06-18 at 00 29 12

**Comment by samsonasik** [comment]
Finally 🎉 🎉 🎉 All checks have passed 🎉 @TomasVotruba it is ready for review.

**Comment by samsonasik** [comment]
rebased, it seems init Scope tweak no longer needed as handled now on `PHPStanNodeScopeResolver` 🎉

**Comment by samsonasik** [comment]
All checks have passed 🎉 @TomasVotruba I think it is ready.

**Comment by samsonasik** [comment]
I will re-check downgrade locally once more

File: full_build.sh [hunk 1]
```diff
@@ -20,6 +20,13 @@ rm -rf vendor
 composer clear-cache
 
 composer install --ansi
+
+# ensure clear cache
+bin/rector --version --clear-cache
+
+# ensure remove cache directory
+php -r 'shell_exec("rm -rf " . sys_get_temp_dir() . "/rector_cached_files");';
```
Comment (samsonasik) [inline full_build.sh:28]: this is really needed to ensure cleared data before run downgrade on `full_build.sh` locally.

**Comment by samsonasik** [comment]
All checks have passed 🎉  @TomasVotruba I think it is ready.

I tested locally once more and it seems working ok 👍 

Screen Shot 2022-06-18 at 13 58 30

**Comment by samsonasik** [comment]
rebased.

**Comment by samsonasik** [comment]
@TomasVotruba I applied performance improvement to avoid loop on `StmtsAwareInterface` instance 👍 

- Only recursive call on `UnreachableStatementNode` OriginalNode and its next Node (yes, next node of `UnreachableStatementNode` is detected as "reachable" while no Scope, so need to be filled with Scope)
- The `Trycatch` catch stmts is special case that next catch can be no Scope as previously catch Throwable already, which next catch is marked as "unreachable".

https://github.com/rectorphp/rector-src/pull/2525/commits/07ce2c3b0b6da60358a1037c462af96279faa21b

I tested locally and the downgrade script is working ok 🎉 🎉 🎉

**Comment by samsonasik** [comment]
@TomasVotruba I applied another performance improvement by set `AttributeKey::IS_UNREACHABLE` to `true` on next Stmt of `UnreachableStatementNode`, so the `UnreachableStmtAnalyzer` will be faster for check previous Stmt that has `AttributeKey::IS_UNREACHABLE` as already filled https://github.com/rectorphp/rector-src/pull/2525/commits/ef09d84d1ed57759de9256d22b66ed1a6350a6e6 🎉 🎉 🎉

**Comment by samsonasik** [comment]
@TomasVotruba I tested locally and it only took 91 seconds to downgrade 🎉 for full running `./full_build.sh` 🎉🎉🎉

```
sh ./full_build.sh  91.16s user 46.65s system 79% cpu
```

**Comment by samsonasik** [comment]
All checks have passed 🎉 @TomasVotruba I think it is ready.

**Comment by TomasVotruba** [comment]
Thanks 👍

Could you squash to fewer commits? My Github is crashing for review 😅

**Comment by samsonasik** [comment]
@TomasVotruba yes, implemented 🎉 

1. start with add failing test cases and move e2e test to rules tests https://github.com/rectorphp/rector-src/pull/2525/commits/47fd9f8dd4e201a424d744bdecdec6c5ec263653
2. remove tweak scope init https://github.com/rectorphp/rector-src/pull/2525/commits/1fb87a3407f05d76998d7cf6677492919ade03bf
3. Apply the fix for parent return/assign from FuncCall no Scope https://github.com/rectorphp/rector-src/pull/2525/commits/7c8346b079c278bd7b4640b404897e14e3e7074c
4. enhancement, ensure cache cleared before run downgrade locally on fullbuild.sh https://github.com/rectorphp/rector-src/pull/2525/commits/e35be5c90d2544956aebb853abe727033ed2e4b5

**Comment by TomasVotruba** [comment]
Looks good, let's ship it :+1:

**Pull request merged**